### PR TITLE
Add support for customizing the Access Token Request

### LINKED ISSUE #5547: Add RestOperations implementation of OAuth2AccessTokenResponseClient
The current (and default) implementation of `OAuth2AccessTokenResponseClient` for the `authorization_code` grant is `NimbusAuthorizationCodeTokenResponseClient`, which leverages the Nimbus HTTP Client internally.

We should provide a new implementation of `OAuth2AccessTokenResponseClient` for the `authorization_code` grant named `DefaultAuthorizationCodeTokenResponseClient` that uses `RestOperations` for requesting to the Token Endpoint.

Additionally, `DefaultAuthorizationCodeTokenResponseClient` should expose `setRestOperations(restOperations)` method to allow user's the ability to supply a pre-configured `RestOperations`, which can provide the user the capability to add custom parameters and/or headers if required.

`DefaultAuthorizationCodeTokenResponseClient` should be the new default in `OAuth2LoginConfigurer` and `OAuth2ClientConfigurer`.

Related #5601

### PR DESCRIPTION
**Pull request by jgrandja**
This PR resolves #5547 and #5466 

The 1st commit pushes most of the code from `DefaultClientCredentialsTokenResponseClient` to an abstract implementation `AbstractOAuth2AccessTokenResponseClient` to allow for reuse in the next commit.

The 2nd commit adds a new implementation of `OAuth2AccessTokenResponseClient` for the `authorization_code` grant `DefaultAuthorizationCodeTokenResponseClient` - extending `AbstractOAuth2AccessTokenResponseClient`.

The 3rd commit makes `DefaultAuthorizationCodeTokenResponseClient` the default client replacing the previous default `NimbusAuthorizationCodeTokenResponseClient`.

### CODE REVIEW

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractOAuth2AccessTokenResponseClient.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+	protected AbstractOAuth2AccessTokenResponseClient() {
+		RestTemplate restTemplate = new RestTemplate();
+		// Disable the ResponseErrorHandler as errors are handled directly within this class
+		restTemplate.setErrorHandler(new NoOpResponseErrorHandler());
+		this.restOperations = restTemplate;
+	}
+
+	@Override
+	public OAuth2AccessTokenResponse getTokenResponse(T authorizationGrantRequest) throws OAuth2AuthenticationException {
```
Comment (gburboz) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractOAuth2AccessTokenResponseClient.java]: Rename parameter `authorizationGrantRequest` to `accessTokenRequest` and mark it `final`

**Comment by rwinch** [comment]
At a high level, I don't really like the idea of the `AbstractOAuth2AccessTokenResponseClient`. It seems to be introducing additional APIs that are already available in RestTemplate. For example, there is a `buildRequestMethod` which is already supported in RestTemplate using a MessageConverter. The conversion of the error response is already possible using a `ResponseErrorHandler`. The parsing of the response is possible using a MessageConverter as well.

I propose the code be refactored to be exposed using the existing APIs within RestTemplate. The DefaultClientCredentialsTokenResponseClient and DefaultAuthorizationCodeTokenResponseClient can then use a RestTemplate with the appropriate API implementations registered by default. User's would then be required to provide a RestOperations implementation that can convert from a request/response to the OAuth2AccessTokenResponseClient. Note that proper setup of the RestTempalte is already required since the AbstractOAuth2AccessTokenResponseClient relies on a NoOpResponseErrorHandler being set to handle the errors.

**Comment by jgrandja** [comment]
@rwinch Ok this makes sense as far as leveraging `HttpMessageConverter` for reading `OAuth2AccessTokenResponse` on a success response and reading `OAuth2Error` using `ResponseErrorHandler` on error response.

Can you clarify:

> there is a buildRequestMethod which is already supported in RestTemplate using a MessageConverter.

I can't seem to find `buildRequestMethod`? Did you mean to leverage `RequestCallback`?

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractOAuth2AuthorizationGrantRequest.java [hunk 2]
```diff
@@ -26,19 +27,34 @@
...
+	}
+
+	/**
+	 * Sub-class constructor.
+	 *
+	 * @since 5.1
+	 * @param authorizationGrantType the authorization grant type
+	 * @param clientRegistration the client registration
+	 */
+	protected AbstractOAuth2AuthorizationGrantRequest(
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractOAuth2AuthorizationGrantRequest.java]: Is this change necessary for this PR?

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+ */
+package org.springframework.security.oauth2.client.endpoint;
+
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.RequestEntity;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.client.ClientHttpResponse;
+import org.springframework.http.converter.FormHttpMessageConverter;
+import org.springframework.security.oauth2.client.endpoint.converter.OAuth2AuthorizationCodeGrantRequestEntityConverter;
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java]: This is a package tangle because a parent package is accessing a subpackage of itself.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+ *
+ * @author Joe Grandja
+ * @since 5.1
+ * @see OAuth2AccessTokenResponseClient
+ * @see OAuth2AuthorizationCodeGrantRequest
+ * @see OAuth2AccessTokenResponse
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc6749#section-4.1.3">Section 4.1.3 Access Token Request (Authorization Code Grant)</a>
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc6749#section-4.1.4">Section 4.1.4 Access Token Response (Authorization Code Grant)</a>
+ */
+public class DefaultAuthorizationCodeTokenResponseClient implements OAuth2AccessTokenResponseClient<OAuth2AuthorizationCodeGrantRequest> {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java]: Most of the methods are already final, let's make the class final

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultClientCredentialsTokenResponseClient.java [hunk 5]
```diff
@@ -151,120 +97,37 @@ public OAuth2AccessTokenResponse getTokenResponse(OAuth2ClientCredentialsGrantRe
...
 	}
 
 	/**
-	 * Sets the {@link RestOperations} used when requesting the access token response.
+	 * Sets the {@link RestOperations} used when requesting the OAuth 2.0 Access Token Response.
 	 *
-	 * @param restOperations the {@link RestOperations} used when requesting the access token response
+	 * @param restOperations the {@link RestOperations} used when requesting the Access Token Response
 	 */
 	public final void setRestOperations(RestOperations restOperations) {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultClientCredentialsTokenResponseClient.java]: How can someone inject the RestOperations and still get valid error handling? I think we may need to make OAuth2ErrorResponseErrorHandler public.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/converter/OAuth2AuthorizationCodeGrantRequestEntityConverter.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		return new RequestEntity<>(formParameters, headers, HttpMethod.POST, uri);
+	}
+
+	/**
+	 * Returns the {@link HttpHeaders} used for the Access Token Request.
+	 *
+	 * @param authorizationCodeGrantRequest the authorization code grant request
+	 * @return the {@link HttpHeaders} used for the Access Token Request
+	 */
+	protected HttpHeaders buildHeaders(OAuth2AuthorizationCodeGrantRequest authorizationCodeGrantRequest) {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/converter/OAuth2AuthorizationCodeGrantRequestEntityConverter.java]: I'd like this to be private. User's that need to customize the headers can do so by delegation to this class and constructing a modified RequestEntity.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/converter/OAuth2AuthorizationCodeGrantRequestEntityConverter.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+		return headers;
+	}
+
+	/**
+	 * Returns a {@link MultiValueMap} of the form parameters used for the Access Token Request body.
+	 *
+	 * @param authorizationCodeGrantRequest the authorization code grant request
+	 * @return a {@link MultiValueMap} of the form parameters used for the Access Token Request body
+	 */
+	protected MultiValueMap<String, String> buildFormParameters(OAuth2AuthorizationCodeGrantRequest authorizationCodeGrantRequest) {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/converter/OAuth2AuthorizationCodeGrantRequestEntityConverter.java]: I'd like this to be private. User's that need to customize the headers can do so by delegation to this class and constructing a modified RequestEntity.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/converter/OAuth2ClientCredentialsGrantRequestEntityConverter.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+		return new RequestEntity<>(formParameters, headers, HttpMethod.POST, uri);
+	}
+
+	/**
+	 * Returns the {@link HttpHeaders} used for the Access Token Request.
+	 *
+	 * @param clientCredentialsGrantRequest the client credentials grant request
+	 * @return the {@link HttpHeaders} used for the Access Token Request
+	 */
+	protected HttpHeaders buildHeaders(OAuth2ClientCredentialsGrantRequest clientCredentialsGrantRequest) {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/converter/OAuth2ClientCredentialsGrantRequestEntityConverter.java]: I'd like this to be private. User's that need to customize the headers can do so by delegation to this class and constructing a modified RequestEntity.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/converter/OAuth2ClientCredentialsGrantRequestEntityConverter.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+		return headers;
+	}
+
+	/**
+	 * Returns a {@link MultiValueMap} of the form parameters used for the Access Token Request body.
+	 *
+	 * @param clientCredentialsGrantRequest the client credentials grant request
+	 * @return a {@link MultiValueMap} of the form parameters used for the Access Token Request body
+	 */
+	protected MultiValueMap<String, String> buildFormParameters(OAuth2ClientCredentialsGrantRequest clientCredentialsGrantRequest) {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/converter/OAuth2ClientCredentialsGrantRequestEntityConverter.java]: I'd like this to be private. User's that need to customize the headers can do so by delegation to this class and constructing a modified RequestEntity.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultClientCredentialsTokenResponseClient.java (hunk 5 above)
Comment (rwinch) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultClientCredentialsTokenResponseClient.java]: We should document what the RestOperations required (i.e. the types of converters it would need, error handling, etc)

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+		Assert.notNull(requestEntityConverter, "requestEntityConverter cannot be null");
+		this.requestEntityConverter = requestEntityConverter;
+	}
+
+	/**
+	 * Sets the {@link RestOperations} used when requesting the OAuth 2.0 Access Token Response.
+	 *
+	 * @param restOperations the {@link RestOperations} used when requesting the Access Token Response
+	 */
+	public final void setRestOperations(RestOperations restOperations) {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java]: Same comments as the other setRestOperations

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java (hunk 3 above)
Comment (rwinch) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java]: i'm wondering if to fix the tangle if we could move `OAuth2AuthorizationCodeGrantRequestEntityConverter` and `OAuth2ClientCredentialsGrantRequestEntityConverter` to `org.springframework.security.oauth2.core.http.converter`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractOAuth2AuthorizationGrantRequest.java (hunk 2 above)
Comment (jgrandja) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractOAuth2AuthorizationGrantRequest.java]: Agreed this change can be removed

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java (hunk 3 above)
Comment (jgrandja) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java]: `OAuth2AuthorizationCodeGrantRequest` and `OAuth2ClientCredentialsGrantRequest` reside in the `client.endpoint` package so we should keep it in the `client` module.

The 2 options I can think of right now are:
`org.springframework.security.oauth2.client.endpoint` (keep in same package)
`org.springframework.security.oauth2.client.converter`

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultClientCredentialsTokenResponseClient.java (hunk 5 above)
Comment (jgrandja) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultClientCredentialsTokenResponseClient.java]: I was thinking of exposing `OAuth2ErrorResponseErrorHandler` but the implementation was pretty easy so decided to hide it.

Either way, I'm fine with exposing it.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java (hunk 3 above)
Comment (rwinch) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java]: Good point, I think given it is associated with RequestEntity, we should consider a variation of what you proposed. What do you think of `org.springframework.security.oauth2.client.http.converter`

Comment (jgrandja) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java]: Agreed that makes sense....let's go with `org.springframework.security.oauth2.client.http.converter`

**Comment by jgrandja** [comment]
@rwinch I've applied the updates as discussed. I've also rebased/squashed to 1 commit and excluded the changes to `DefaultClientCredentialsTokenResponseClient`. After this is merged, I'll resolve #5735

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+ */
+package org.springframework.security.oauth2.client.endpoint;
+
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.http.RequestEntity;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.converter.FormHttpMessageConverter;
+import org.springframework.http.converter.HttpMessageConverter;
+import org.springframework.security.oauth2.client.http.OAuth2ErrorResponseErrorHandler;
+import org.springframework.security.oauth2.client.http.converter.OAuth2AuthorizationCodeGrantRequestEntityConverter;
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/DefaultAuthorizationCodeTokenResponseClient.java]: Unfortunately this is another package tangle. The `client.endpoint` package refers to `http.converter` above and `http.converter` refers to `client.endpoint`

**Comment by jgrandja** [comment]
`DefaultAuthorizationCodeTokenResponseClient` and `OAuth2AuthorizationCodeGrantRequest` reside in the `client.endpoint` so I moved `OAuth2AuthorizationCodeGrantRequestEntityConverter` to the same package as this made the most sense rather than it's own `http.converter` package.

`DefaultAuthorizationCodeTokenResponseClient` **uses** `OAuth2AuthorizationCodeGrantRequestEntityConverter` which **uses** `OAuth2AuthorizationCodeGrantRequest`.

**Comment by rwinch** [comment]
Thanks please squash and rebase off master as there have been some recent checkstyle changes. Once that is done and the build passes, please feel free to merge.

**Comment by jgrandja** [comment]
Thanks @rwinch. This is now in master.
